syntax = "proto3";

package user_service;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";


message UserRequest {
  string name = 1;
  int32 training_length = 2;
  string password = 3;
  string email = 4;
  string telegram_user_id = 5;
}

message UserIdRequest {
  string request_id = 1;  // UUID
}

message UserNameRequest {
  string user_name = 1;
}

message UserTrainingLengthRequest {
  string user_name = 1;
  int32 training_length = 2;
}

message UserCreateTelegramDTORequest {
  string id = 1;  // UUID
  string user_name = 2;
  int32 training_length = 3;
  string password = 4;
  string email = 5;
  string telegram_user_id = 6;
}

message CheckResponse {
  bool result = 1;
}

message UserIdResponse {
  string user_id = 1;  // UUID
}

message UserCreateFullDTOResponse {
    string id = 1; // UUID
    string user_name = 2;
    int32 training_length = 3;
    google.protobuf.Timestamp created_at = 4;
    google.protobuf.Timestamp updated_at = 5;
}


service UserServiceGRPC {
  rpc get_user_by_id (UserIdRequest) returns (UserCreateFullDTOResponse);
  rpc get_user_by_name (UserNameRequest) returns (UserCreateFullDTOResponse);
  rpc get_user_id_by_name (UserNameRequest) returns (UserIdResponse);
  rpc update_user_training_length (UserTrainingLengthRequest) returns (google.protobuf.Empty);
  rpc is_user_created (UserNameRequest) returns (CheckResponse);
  rpc create_user_by_DTO (UserCreateTelegramDTORequest) returns (google.protobuf.Empty);
  rpc create_user (UserRequest) returns (google.protobuf.Empty);
}